packages/multichain/src/scope/validation.ts

```diff
@@ -0,0 +1,129 @@
+import { isCaipReference } from '@metamask/utils';
+
+import type {
+  ExternalScopeString,
+  ExternalScopeObject,
+  ExternalScopesObject,
+} from './types';
+import { parseScopeString } from './types';
+
+/**
+ * Validates a scope object according to the [CAIP-217](https://chainagnostic.org/CAIPs/caip-217) spec.
+ * @param scopeString - The scope string to validate.
+ * @param scopeObject - The scope object to validate.
+ * @returns A boolean indicating if the scope object is valid according to the [CAIP-217](https://chainagnostic.org/CAIPs/caip-217) spec.
+ */
+export const isValidScope = (
+  scopeString: ExternalScopeString,
+  scopeObject: ExternalScopeObject,
+): boolean => {
+  const { namespace, reference } = parseScopeString(scopeString);
+
+  // Namespace is required
+  if (!namespace) {
+    return false;
+  }
+
+  const {
+    references,
+    methods,
+    notifications,
+    accounts,
+    rpcDocuments,
+    rpcEndpoints,
+    ...extraProperties
+  } = scopeObject;
+
+  // Methods and notifications are required
+  if (!methods || !notifications) {
+    return false;
+  }
+
+  // For namespaces other than 'wallet', either reference or non-empty references array must be present
+  if (
+    namespace !== 'wallet' &&
+    !reference &&
+    (!references || references.length === 0)
+  ) {
+    return false;
+  }
+
+  // If references are present, reference must be absent and all references must be valid
+  if (references) {
+    if (reference && references.length > 0) {
```

Comment:
Should we allow an empty `references` entry if `reference` is defined? The spec seems to suggest that it would be disallowed.

But if we did allow it, it would just get removed by our normalization I think, so it would be harmless. Still maybe best to disallow for spec compliance, not sure.

```suggestion
    if (reference) {
```

Reply:
Regardless which way we go here, we should cover this in a test as well

Reply:
probably cleaner and fine to just enforce it as one or the other

Reply:
done here https://github.com/MetaMask/core/pull/4784/commits/e5762958f52c21a6d44fb46ae262601e0033951f